cpp/validate_harvested_records.cc

```diff
@@ -39,274 +40,195 @@
+    std::set<char> found_subfield_codes;
+    for (const auto &subfield : subfields)
+        found_subfield_codes.emplace(subfield.code_);
 
-GeneralInfo GeneralInfo::Combine(const GeneralInfo &lhs, const GeneralInfo &rhs) {
-    auto lhs_iter(lhs.begin());
-    auto rhs_iter(rhs.begin());
-
-    GeneralInfo combined_info;
-    while (lhs_iter != lhs.end() or rhs_iter != rhs.end()) {
-        if (lhs_iter == lhs.end()) {
-            combined_info.addField(*rhs_iter);
-            ++rhs_iter;
-        } else if (rhs_iter == rhs.end()) {
-            combined_info.addField(*lhs_iter);
-            ++lhs_iter;
-        } else {
-            const int compare_result(StringUtil::AlphaWordCompare(lhs_iter->name_, rhs_iter->name_));
-            if (compare_result == 0) {
-                // if present on both sides, rhs wins!
-                combined_info.addField(*rhs_iter);
-                ++lhs_iter;
-                ++rhs_iter;
-            } else if (compare_result < 0) {
-                combined_info.addField(*lhs_iter);
-                ++lhs_iter;
-            } else if (compare_result > 0) {
-                combined_info.addField(*rhs_iter);
-                ++rhs_iter;
-            }
+    for (const auto &subfield_code_and_field_presence : subfield_code_to_field_presence_map_) {
+        const auto iter(found_subfield_codes.find(subfield_code_and_field_presence.first));
+        if (iter == found_subfield_codes.end() and subfield_code_and_field_presence.second == ALWAYS) {
+            if (not reason_for_being_invalid->empty())
+                reason_for_being_invalid->append("; ");
+            reason_for_being_invalid->append("required subfield " + std::string(1, subfield_code_and_field_presence.first)
+                                             + " is missing");
         }
     }
-
-    return combined_info;
 }
 
 
-class JournalInfo : public GeneralInfo {
-    std::string zeder_id_;
-    std::string zeder_instance_;
-    bool not_in_database_yet_;
+class Validator {
 public:
-    JournalInfo(const std::string &zeder_id, const std::string &zeder_instance,
-                const bool not_in_database_yet): zeder_id_(zeder_id), zeder_instance_(zeder_instance),
-                not_in_database_yet_(not_in_database_yet) { }
-    JournalInfo() = default;
-    JournalInfo(const JournalInfo &rhs) = default;
-
-    const std::string &getZederId() const { return zeder_id_; }
-    const std::string &getZederInstance() const { return zeder_instance_; }
-    bool isInDatabase() const { return not not_in_database_yet_; }
+    ~Validator() {}
+
+    /** \return True if we found rules for all subfields in "field" o/w false.
+     *  \note   If a rule violation was found, "reason_for_being_invalid" w/ be non-empty after the call and
+     *          we will return true.
+     */
+    virtual bool foundRuleMatch(const std::string &journal_id, const MARC::Record::Field &field,
```

Comment:
Maybe we can find a better name for the function. The name suggests that this function only checks for rules that match against the field, but it's more about validating a Field against these rules, so i guess `validateField` would be better

Reply:
Regarding the return value, i would also suggest that a `validate` function might be clearer. A field should be treated as valid (true) if for each subfield, either the rule is fulfilled or there is no rule. Only if at least 1 existing rule is violated, the validation should fail.

Reply:
ah, now i understand why it's needed this way without validation, sorry for bothering.